### DISCUSSION TITLE
How to set access operation.query/filter.query based on requested item data?

### DISCUSSION THREAD

**Discussion by wysher** [keystonejs/keystone#8446] [Questions]
Is there a way to get currently requested item data on access.operation.query / access.filter.query?

This seems like common thing to do and should be possible (I thought of few other examples where I'd need that), but I cannot find any solution.

I have Course model with availableAt (timestamp field) and relationship to CourseModule.
CourseModule has availableAfterDays (integer field).

```js
export const Course = list({
    fields: {
        availableAt: timestamp({}),

        modules: relationship({
            ref: 'Module.courses',
            many: true,
        }),
    },
});
```

```js
export const CourseModule = list({
    fields: {
        availableAfterDays: integer({
            defaultValue: 0,
        }),

        courses: relationship({
            ref: 'Course.modules',
            many: true,
        }),
    },
});

```

Each CourseModule may be accessed when Date.now() is passed Course.availableAt + CourseModule.availableAfterDays

And now for CourseModule I need to disable access or filter out when it's unavailable. But I cannot get requested item data to make calculation, like this:

```js
export const CourseModule = list({
    access: {
        filter: {
            query: ({ item }) => { // I know there's no item here, but how to get it?
                const moduleAvailableAt = new Date(item.course.availableAt);
                moduleAvailableAt.setDate(moduleAvailableAt.getDate() + item.availableAfterDays);
                return {
                    courses: {
                        some: {
                            availableAt: {
                                lt: moduleAvailableAt,
                            },
                        },
                    },
                };
            },
        },
    },
};
```

Is there any way to get this item? I tried looking through context.req/res but trying to use await args.context.db.Module.findMany(req.body.variables) causes heap out of memory error (reasonable).

Or there's only extendGraphqlSchema / filter items on frontend?

**Comment by wysher** [comment]
I found solution which seems to works for me, but still I'm not sure if it's correct or there's better ways to filter using current item.

I need to admit, that now I see why there's no item in filters - because filter called before request is even called. But maybe there are plans of extending filter abilities in the future?

So here's what I tried (maybe someone will find this helpful):

1. using context.sudo() will not cause `heap out of memory` error, because request is not recursively called, which happens without sudo. So... I just await for CourseModule using context.req.body.variables, filter calculated data and return ids, like so:

```js
export const CourseModule = list({
    access: {
        filter: {
            query: async (args) => {
                const modules = await args.context?.sudo().query.CourseModule.findMany({
                    ...args.context.req?.body.variables,
                    query: 'availableAfterDays course { availableAt  }',
                });

                const now = new Date();

                const ids = modules?.filter(({ course, availableAfterDays }) => {
                    const moduleAvailableAt = new Date(course.availableAt);
                    moduleAvailableAt.setDate(moduleAvailableAt.getDate() + availableAfterDays);
                    return moduleAvailableAt < now;
                });

                return {
                    id_in: ids?.map(({ id }) => id),
                };
            },
        },
    },
});
```

One downside of this approach is that if I have nested to CourseModule model, like CourseLesson it should be calculated a little bit differently, because `args.context.req?.body.variables` will have variables for CourseLessons, not CourseModule. There's listKey, so not that bad.

---

1.5. I thought of using virtual field, but you can use virtual field for filtering. Which seems reasonable.

---

2. Second approach will work as well, but I think it shouldn't be done. Not only because it's daunting, but it's also error prone. It requires adding "pseudo-virtual" field which developer needs to sync using hooks.resolveInput for CourseModule AND Course. Code for CourseModule looks like this
```js
export const CourseModule = baseList({
    access: {
        operation: {
            create: rules.isAdmin,
            query: rules.isSignedIn,
            update: rules.isAdmin,
            delete: rules.isAdmin,
        },
        filter: {
            query: (args) =>
                rules.isAdmin(args) || {
                    availableAt: {
                        lt: new Date(),
                    },
                },
        },
    },
    fields: {
        availableAfterDays: integer({}),

        availableAt: timestamp({
            defaultValue: {
                kind: 'now',
            },
        }),

        course: relationship({
            ref: 'Course.modules',
            many: false,
        }),
    },
    hooks: {
        async resolveInput({ context, item, resolvedData }) {
            if (resolvedData.availableAfterDays === item?.availableAfterDays) return resolvedData;

            const nextAvailableAfterDays =
                resolvedData.availableAfterDays || item?.availableAfterDays;
            let nextCourse;

            // yeah, we need to keep track of connecting refs
            if (resolvedData.course?.connect?.id) {
                const course = await context.query.Course.findOne({
                    where: { id: resolvedData.course.connect.id },
                    query: 'availableAt',
                });

                nextCourse = course;
            } else {
                const data = await context.query.CourseModule.findOne({
                    where: { id: item?.id },
                    query: 'course { availableAt }',
                });

                nextCourse = data.course;
            }

            const nextAvailableAt = new Date(nextCourse.availableAt);
            nextAvailableAt.setDate(nextAvailableAt.getDate() + nextAvailableAfterDays);

            resolvedData.availableAt = nextAvailableAt;
            return resolvedData;
        },
    },
});
```

As mentioned before, similar hook should be added for Course to update CourseModule if course.availableAt changes or models were disconnected.

---

3. I also thought of extendGraphqlSchema, but I want to have proper access control in AdminUI, and AFAIK you cannot update existing query (which is used by AdminUI).

---

That's it. If someone knows better way and can share I would appreciate :)

I also think this problem with solution, should be described in documentation.

**Reply by marekryb** [reply]
Hi,

In general scope of things, I think this comes down to your data model. With your current many-to-many relation, every time you query N CourseModule you need to make N*M db queries (where M is number of related courses) just for access control. With big enough data set this is unsustainable. Probably you could write some clever raw SQL query that can do what you desire (complex sum of fields in two tables inside WHERE), but that would be very hard to couple with keystone's access control.

How about splitting your many-to-many relation to two one-to-many? Something like
Course -> CourseModuleItem <- CourseModule

Whenever your create Course, you would create also CourseModuleItem(s) which have exact timestamp indicating whether the can be shown to the user or not (availableAt + availableAfterDays). Then you can use standard filters which would have way better performance as they just add additional WHERE clause.

**Reply by wysher** [reply]
Thank you for that. I didn't think about performance issues with my data model. As I was working more at project I expanded it much further with more "nested" relations, so I should probably consider this ASAP.

Correct me if I'm wrong. You suggest to do something similar I have in second approach. But i tried to update item itself with resolveInput hook on course and module upsert. Instead I should use resolveInput hook to update CourseModuleItem whenever course or module changes? This sounds a little bit easier.

---
Not related to question, but maybe you may help me a little bit with this, assuming I use your approach.
Currently I have Offers > Products > Courses > Modules > Lessons. Plus Purchase connected to Offers. 

AvailableAfterDays is at Lesson model. Lessons access depends as well on Offers (startAt/endAt) and Purchase (with expiresAt). Should I have sth like
Lesson -> PurchasedOfferLesson <- Offer & Purchase (so in general three one to many relations) with calculated dates describing access?

Same goes with Modules and Course, which access depends on Lesson access... so I need another relation. Maybe for each I need to have one-to-one with PurchasedOfferLesson and "just" keep PurchasedOfferLesson in sync on each course/module/offer/purchase change?

**Reply by marekryb** [reply]
>Correct me if I'm wrong. You suggest to do something similar I have in second approach. But i tried to update item itself with resolveInput hook on course and module upsert. Instead I should use resolveInput hook to update CourseModuleItem whenever course or module changes? This sounds a little bit easier.

My initial thought was like this, yes. But now I noticed that in your 2nd approach `Course` and `CourseModule` relation is now changed to one-to-many already. `resolveInput` that updates `availableAt` on `CourseModule` should work fine  - like your 2nd approach.
You can simplify it if you can make (IMO reasonable) assumption that module cannot change course - just disable updating relationship entirely or by non-admin.
`CourseModuleItem` approach might also be fine, but it is additional maintenance headache and doesn't seem necessary.

> AvailableAfterDays is at Lesson model. Lessons access depends as well on Offers (startAt/endAt) and Purchase (with expiresAt). Should I have sth like
Lesson -> PurchasedOfferLesson <- Offer & Purchase (so in general three one to many relations) with calculated dates describing access?

Without fully understanding your requirements and model it is hard to give advice. If your `Purchase` is related to `Offer` then how would `Lesson` know whether user has access or not? Go all the way up?
```typescript
{ 
  module: { 
    course: { 
      products: {
        some: {
          offers: {
            some: {
              startAt: { gte: now }, 
              endAt: { lt: now },
              purchases: { some: { 
                user: { id: { equals: args.session?.itemId }, 
                expiresAt: { gt: now } 
              } } } 
            }
          } 
        }
      } 
    } 
  } 
}
```
Definitely not pretty but might work. Would need to check query performance though.

Making additional tables like `PurchasedLesson` / `PurchasedCourse` etc. might improve this since you would have to go only one level, but it is additional overhead and something to maintain.

Probably there is some better way to solve this, but I cannot think of one right now.
